components/client/src/test/unit/java/com/hotels/styx/client/StyxHostHttpClientTest.java

```diff
@@ -15,55 +15,202 @@
+
+        // The StepVerifier consumes the response event and then unsubscribes
+        // from the response observable.
+        StepVerifier.create(hostClient.sendRequest(request))
+                .consumeNextWith(transformedResponse::set)
+                .verifyComplete();
+
+        // The response is still alive...
+        verify(pool, never()).returnConnection(any(Connection.class));
+        verify(pool, never()).closeConnection(any(Connection.class));
+
+        // ... until response body is consumed:
+        transformedResponse.get().consume();
+
+        // Finally, the connection is returned after the response body is fully consumed:
+        verify(pool).returnConnection(any(Connection.class));
+    }
+
+    @Test
+    public void releasesIfRequestIsCancelledBeforeHeaders() {
+        Connection connection = mockConnection(PublishSubject.create());
+        ConnectionPool pool = mockPool(connection);
+
+        StyxHostHttpClient hostClient = new StyxHostHttpClient(pool);
+
+        StepVerifier.create(hostClient.sendRequest(request), 0)
+                .expectNextCount(0)
+                .thenCancel()
+                .verify();
+
+        verify(pool).closeConnection(any(Connection.class));
+    }
+
+    @Test
+    public void ignoresResponseObservableErrorsAfterHeaders() {
+        Connection connection = mockConnection(responseProvider);
+        ConnectionPool pool = mockPool(connection);
+        AtomicReference<LiveHttpResponse> newResponse = new AtomicReference<>();
+
+        StyxHostHttpClient hostClient = new StyxHostHttpClient(pool);
+
+        StepVerifier.create(hostClient.sendRequest(request))
+                .then(() -> {
+                    responseProvider.onNext(response);
+                    responseProvider.onError(new RuntimeException("oh dear ..."));
+                })
+                .consumeNextWith(newResponse::set)
+                .expectError()
+                .verify();
+
+        newResponse.get().consume();
 
-        when(transport.send(any(LiveHttpRequest.class), any(ConnectionPool.class))).thenReturn(transaction);
-        when(transaction.response()).thenReturn(Observable.just(response().build()));
+        verify(pool).returnConnection(any(Connection.class));
+    }
+
+    @Test
+    public void terminatesConnectionWhenResponseObservableCompletesWithoutHeaders() {
+        // A connection that yields no response:
+        Connection connection = mockConnection(Observable.empty());
+        ConnectionPool pool = mockPool(connection);
+
+        StyxHostHttpClient hostClient = new StyxHostHttpClient(pool);
+
+        StepVerifier.create(hostClient.sendRequest(request))
+                .expectNextCount(0)
+                .expectComplete()
+                .log()
+                .verify();
+
+        verify(pool).closeConnection(any(Connection.class));
+    }
+
+    @Test
+    public void releasesConnectionWhenRequestFailsBeforeHeaders() {
```

Comment:
Looks like this might be a mistake?
```suggestion
    public void releasesConnectionWhenResponseFailsBeforeHeaders() {
```